monitoring/ping.go

```diff
@@ -0,0 +1,314 @@
+			continue
+		}
+
+		rttNanoSec, err := calculateRTT(client, c.self, node)
+		if err != nil {
+			return trace.Wrap(err)
+		}
+
+		_, err = c.saveLatencyStats(rttNanoSec, node)
+		if err != nil {
+			return trace.Wrap(err)
+		}
+
+		latencyInterface, exists := c.latencyStats.Get(node.Name)
+		if !exists {
+			return trace.NotFound("latency for %s not found", node.Name)
+		}
+		latency, ok := latencyInterface.([]int64)
+		if !ok {
+			return trace.BadParameter("expected latency for %s to be []int64 got %T", c.serfMemberName, latency)
+		}
+		latencyHistogram, err := c.buildLatencyHistogram(node.Name)
+		if err != nil {
+			return trace.Wrap(err)
+		}
+
+		c.logger.Debugf("%s <-ping-> %s = %dns [latest]", c.self.Name, node.Name, rttNanoSec)
+		c.logger.Debugf("%s <-ping-> %s = %dns [%.2f percentile]",
+			c.self.Name, node.Name,
+			latencyHistogram.ValueAtQuantile(latencyQuantile),
+			latencyQuantile)
+
+		latencyPercentile := latencyHistogram.ValueAtQuantile(latencyQuantile)
+		if latencyPercentile >= latencyThreshold.Nanoseconds() {
+			c.logger.Warningf("%s <-ping-> %s = slow ping detected. Value %dns over threshold %s (%dns)",
+				c.self.Name, node.Name, latencyPercentile,
+				latencyThreshold.String(), latencyThreshold.Nanoseconds())
+		} else {
+			c.logger.Debugf("%s <-ping-> %s = ping okay. Value %dns within threshold %s (%dns)",
+				c.self.Name, node.Name, latencyPercentile,
+				latencyThreshold.String(), latencyThreshold.Nanoseconds())
+		}
+	}
+
+	return nil
+}
+
+// buildLatencyHistogram converts the []int64 of latencies in a HDRHistrogram
+func (c *pingChecker) buildLatencyHistogram(nodeName string) (latencyHDR *hdrhistogram.Histogram, err error) {
+	latencyHDR = hdrhistogram.New(pingMinimum.Nanoseconds(),
+		pingMaximum.Nanoseconds(), pingSignificantFigures)
+
+	latency, exists := c.latencyStats.Get(nodeName)
+	if !exists {
+		return nil, trace.NotFound("latency for %s not found", nodeName)
+	}
+	latencySlice, ok := latency.([]int64)
+	if !ok {
+		return nil, trace.BadParameter("couldn't parse node latency as []int64 for %s", nodeName)
+	}
+
+	for _, v := range latencySlice {
+		err := latencyHDR.RecordValue(v)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+	}
+
+	return latencyHDR, nil
+}
+
+// saveLatencyStats is used to store ping values in HDR Histograms in memory
+func (c *pingChecker) saveLatencyStats(pingLatency int64, node serf.Member) (latencies []int64, err error) {
+	c.mux.Lock()
+	defer c.mux.Unlock()
+
+	if value, exists := c.latencyStats.Get(node.Name); exists {
+		var ok bool
+		if latencies, ok = value.([]int64); !ok {
+			return nil, trace.BadParameter("couldn't parse node latency as []int64 on %s", c.serfMemberName)
+		}
+	}
+
+	if len(latencies) >= slidingWindowSize {
+		// pop oldest value to make room for the new one
+		copy(latencies, latencies[1:])
+		// keep the slice within the sliding window size
+		latencies = latencies[:slidingWindowSize-1]
```

Comment:
Why `-1`? You're losing the last element. Also, why do you need "copy" first? I would simplify all this to just:

```go
if len(latencies) >= slidingWindowSize {
  latencies = latencies[1:slidingWindowSize] // or event just latencies = latencies[1:] since it's never gonna exceed it
}
```

Reply:
slidingWindowSize needs to have -1 because another element will be appended a few lines below as the latest/newest element.

Removed copy, it was there from a proposal that @a-palchikov did but I guess it's redundant now.